README.md
# nerves_initramfs

[![CircleCI](https://circleci.com/gh/nerves-project/nerves_initramfs.svg?style=svg)](https://circleci.com/gh/nerves-project/nerves_initramfs)

This project creates a tiny
[initramfs](https://en.wikipedia.org/wiki/Initial_ramdisk) that supports
cross-platform failure recovery and non-trivial root filesystem setups on
devices using Nerves.

The way it works is this:

1. The device's bootloader (U-Boot, Barebox, RPi, grub, etc.) boots the
   platform-specific Linux kernel and the initramfs from this project
2. The `init` process provided by this project handles any recovery tasks, and
   then mounts the real root filesystem, and runs the real `init` process
   (`erlinit` for Nerves)

This project is focused on recovery and root filesystem mounts. No
other functionality is likely to be added to this project to keep it small,
fast, and easy to audit.

Failure recovery at this level means that if a Nerves device fails
catastrophically after the Linux kernel starts, code here can revert devices to
previous firmware or start recovery mechanisms. This project does not protect
against Linux kernel initialization crashes. This is a tradeoff to avoid
maintaining an implementation for every bootloader where this logic is time
consuming to verify. You can, of course, transition away from this code as
needed.

Root filesystem mount functionality is handled here as well since the initramfs
is typically where this feature lives for when the root filesystem is encrypted.
Additionally, if the root filesystem fails to decrypt, the recovery mechanisms
described previously can be triggered.

## Boot configuration

`nerves_initramfs` is configured via rules in a file called
`/nerves_initramfs.conf` in the initramfs. Be aware that `nerves_initramfs` will
delete that file when it cleans up before switching to the true root filesystem.

The `nerves_initramfs.conf` file uses a simple language for specifying rules and
setting variables. Rules are composed of a condition and a list of actions .
Here's an example:

```config
!fw_validated && fw_booted -> { fwup_revert(); reboot(); }
```

Real configuration files contain rules to handle fallback logic or set up root
filesystem mounts that Linux wouldn't be able to do without help.

### Variables

Variables can be defined and used as needed like other languages. The following
variables have special uses:

Variable           | Description
-------------------|-------------
rootfs.fstype      | Root filesystem time. Defaults to "squashfs"
rootfs.path        | Root filesystem path or spec. Defaults to "/dev/mmcblk0p2"
rootfs.encrypted   | True if the filesystem is encrypted. Defaults to `false`
rootfs.cipher      | The cipher used to encrypt the filesystem. For example, "aes-cbc-plain"
rootfs.secret      | The secret key as hex digits
uboot_env.path     | The location for U-Boot environment data. Defaults to "/dev/mmcblk0"
uboot_env.loaded   | True if the U-Boot environment block has been loaded.
uboot_env.modified | True if something has modified the U-Boot block and it differs from what's on disk
uboot_env.start    | The block offset of the U-Boot environment. (512 byte blocks)
uboot_env.count    | The number of blocks in the environment. Defaults to 256.
run_repl           | True to run a REPL before booting. This is useful for debug. Defaults to `false`

Variables can be overridden using the Linux commandline. See your platform's
bootloader documentation for how to pass options to Linux. At the end of the
commandline, add a `--` and then add `variable` and `variable=value` strings.
The `--` will stop the Linux kernel from processing parameters so make sure that
everything intended for Linux is on the left side and everything for
`nerves_initramfs` is on the right side.

For example, the following Linux commandline sets the Linux console and then
passes a root filesystem path to `nerves_initramfs` and starts a repl.

```text
console=tty1 -- rootfs.path=/dev/sda2 run_repl
```

### Functions

It's also possible to call built-in functions:

Function           | Description
-------------------|-------------
blkid()            | Print out information about all block devices
cmd()              | Run an external program. The first argument is the path to the program, the next is the first argument, and so on.
env()              | Print out all loaded U-Boot variables
fwup_revert()      | Run fwup with the appropriate parameters to revert to the previous firmware. Reboots on success.
getenv(key)        | Get the value of a U-Boot variable
help()             | Print out help when running in the REPL
print(...)         | Print one or more strings and variables
loadenv()          | Load a U-Boot environment block. Set up `uboot_env.path`, `uboot_env.start` and `uboot_env.count` first.
ls()               | List files a directory
poweroff()         | Power off the device
readfile(path)     | Read a file (truncates long files)
reboot()           | Reset the device
saveenv()          | Save all U-Boot variables back to storage
setenv(key, value) | Set a U-Boot variable. It is not saved until you call `saveenv()`
sleep(timeout)     | Wait for the specified milliseconds
vars()             | Print out all known variables and their values

### Block device specifications

Block devices, such as `/dev/sda2`, are either specified as absolute paths or
using the Linux match syntax. Currently only `PARTUUID` is supported for
identifying a partition by its UUID. Use the `blkid` function or the `blkid`
commandline utility in Linux to list information about block devices.

If you are only using one storage device, using absolute paths to block devices
is fine. If you have more than one storage device, Linux sometimes can enumerate
them in a different order so `/dev/sda` could be `/dev/sdb` sometimes. The way
around this is to identify devices by UUID.

## Building

Users should prefer to use pre-built releases. To build your own, you will need
to use Linux and install the [Buildroot required
packages](https://buildroot.org/downloads/manual/manual.html#requirement).

Change to the `builder` directory and run `./build-all.sh` to build for all
platforms.

If you're only building for one platform, run the following:

```sh
./create-build.sh configs/nerves_initramfs_arm_defconfig o/arm
cd o/arm
make
```

The `nerves_initramfs` images will be in the `images` directory. You can also run `make
menuconfig` to enable other applications and libraries that may be useful for
your particular setup.

## Linux kernel configuration

The following strings must be in your kernel configuration:

* `CONFIG_BLK_DEV_RAM=y`
* `CONFIG_BLK_DEV_INITRD=y`
* `CONFIG_RD_<compression>=y` - gzip compression is enabled by default, but if
  you want lz4, for example, you'll need to enable it.

To mount encrypted filesystems, you'll need these additional configuration strings:

* `CONFIG_BLK_DEV_LOOP=y`
* `CONFIG_MD=y`
* `CONFIG_BLK_DEV_DM=y`
* `CONFIG_DM_CRYPT=y` - Only `dm-crypt` is supported. `cryptoloop` and
  `loop-aes` are not supported.
* `CONFIG_CRYPTO_USER=y`
* `CONFIG_CRYPTO_USER_API_HASH=y`
* `CONFIG_CRYPTO_USER_API_SKCIPHER=y`
* `CONFIG_CRYPTO_AES=y` - Make sure to enable the cryptographic algorithms that
  you're using. Hardware acceleration options may exist as well.

## Preparing files for boot

`initramfs` requires a single file in `cpio` format. If you have multiple files, they
need to be converted to `cpio` format and concatenated together with the build.

The compressed build files are already in cpio format, but you will need to convert
your `nerves_initramfs.confg` and any other files (such as `revert.fw`) then
concatenate them together before using on boot. The helper script `file-to-cpio.sh`
was created for this and is included with the release so it can be used as needed.

Your implementation may differ, but a simple usage example is:

```sh
$ file-to-cpio.sh nerves_initramfs.conf nerves_initramfs.conf.cpio
$ file-to-cpio.sh revert.fw revert.fw.cpio

# Then concatenate files together
# The build .gz/.xz is already cpio format
$ cat nerves_initramfs_arm.gz nerves_initramfs.conf.cpio revert.fw.cpio > nerves_initramfs
$ cp nerves_initramfs /path/to/your/boot/partition
```

## Raspberry Pi configuration

The Raspberry Pi's bootloader supports loading `initramfs` images off the boot
partition that contains the Linux kernel. Since gzip compression is pretty much
always enabled in Linux kernels, the `nerves_initramfs_arm.gz` is safe to use.

First, follow the steps in [**Preparing files for boot**](#preparing-files-for-boot),
copy the `nerves_initramfs` file to the boot partition (in Nerves, you can do this
manually or edit the `fwup.conf` to automate prepping and copying the files),
then add this line to `config.txt`:

```config
initramfs nerves_initramfs followkernel
```

See the [official config.txt boot
documentation](https://www.raspberrypi.org/documentation/configuration/config-txt/boot.md)
for more information.

## U-boot configuration

## Grub configuration

## Mounting an encrypted file system

This project mounts encrypted file systems by using the Linux kernel's
`dm-crypt` module. As such, the web contains a lot of information on this
feature. The main difference, though, is that devices using this project are
expected to be unattended. In other words, a human can not enter a password when
the device boots. This ends up simplifying the options that can be used, but the
handling of the secret can become a major issue.

Storing the secret securely depends on your hardware and your needs. It is
expected that some projects will need to fork this repository to access their
secrets. The examples shown here are meant to be informative on the process, but
they are not secure. They are probably more accurately referred to as ways to
obfuscate the root filesystem.

Here's the general idea:

1. Create your root filesystem as normal
2. Write the filesystem to the destination media. If using a MicroSD card, this
   could be done by mounting it on a Linux PC using `dm-crypt` and writing the
   root filesystem via that. `fwup` also supports encryption.
3. Make sure that your device's Linux kernel supports the chosen encryption
   algorithms
4. Configure this project with how to find or generate the secret key

Here's an example configuration file with the secret key stored as plain text:

```config
rootfs.encrypted = true
rootfs.cipher = "aes-cbc-plain"
rootfs.secret = "8e9c0780fd7f5d00c18a30812fe960cfce71f6074dd9cded6aab2897568cc856"
```

This is illustrative, but obviously quite insecure. The current route to
obtaining the secret key is to edit the C code to this project to integrate it
with platform-specific way of keeping or hiding secrets. It is hoped that
alternatives can be shared in the future.


CHANGELOG.md
# Changelog

## v0.5.0

* New features
  * Added nerves_initramfs support for `aarch64` targets.
  * buildroot 2020.11.1
  * fwup v1.8.3

## v0.4.0

* New features
  * Add `file-to-cpio.sh` helper script to release so that systems using
    `nerves_initramfs` don't need to maintain a copy.
  * fwup v1.7.0

## v0.3.3

* New features
  * Change Linux command line override format to actually work. New format is
    to pass `--` and then `key` or `key=value` parameters. For example,
    `-- rootfs.path=/dev/sda2 run_repl`
  * Create /dev/rootdisk symlinks - These are used various places to generically
    refer to the boot block device. Creating them here is more reliable than
    other locations since the symlinks are based on the nerves_initramfs
    configuration rather than trying to derive them from mount information.
  * Buildroot 2020.02
  * fwup v1.5.2 - footprint reductions

* Bug fixes
  * Fix block calculation for call to device mapper

## v0.3.2

* New features
  * Add a readfile() function

* Bug fixes
  * Fix static build of init

## v0.3.1

* New features
  * Print out version on start
  * Support debug builds. Debug initramfs cpio images can be concatenated onto
    the ones in use to "turn on" debug prints.
  * Rename info() to print() since that's what the function does

* Bug fixes
  * Fix bug where uboot environment reads would not wait for the target disk to
    enumerate
  * Fix issue where UUIDs were case sensitive on the hex digits

## v0.3.0

* New features
  * Support probing disks by UUID (as well as partitions)
  * New commands: ls(), sleep(), reboot(), poweroff(), fwup_revert()

* Bug fixes
  * Fix U-Boot environment block reading/writing
  * Clean up all files before switching the rootfs
  * Various script language fixes

## v0.2.0

Add GPT support and include fwup so that it's possible to revert images.

## v0.1.2

Create one tar file that has all of the release binaries.

## v0.1.1

Rename project and various fixes to the language.

## v0.1.0

Initial release


src/script.c
#include "script.h"
#include "util.h"
#include "parser.tab.h"
#include "block_device.h"
#include "cmd.h"

#include <dirent.h>
#include <fcntl.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>

#include <linux/reboot.h>
#include <sys/reboot.h>

#define HEAP_SIZE 16536

static char *heap = NULL;
static int heap_index = 0;
static struct term *variables = NULL;

/**
 * This can only be called between parsing statements
 * since we don't track references in bison.
 */
void term_gc_heap()
{
    struct term *old_variables;
    char *old_heap;

    if (!heap) {
        old_variables = NULL;
        old_heap = NULL;
    } else {
        old_heap = heap;
        old_variables = variables;
    }

    heap = malloc(HEAP_SIZE);
    memset(heap, 0, HEAP_SIZE);

    heap_index = 0;
    variables = NULL;

    while (old_variables) {
        set_variable(old_variables->var.name, term_dupe(old_variables->var.value));
        old_variables = old_variables->next;
    }

    if (old_heap)
        free(old_heap);
}

static void *alloc_heap(size_t num_bytes)
{
    if (heap_index + num_bytes > HEAP_SIZE)
        fatal("Program too large. Out of heap");

    void *addr = &heap[heap_index];
    heap_index = (heap_index + num_bytes + 7) & ~7;
    return addr;
}

static char *alloc_string(const char *str)
{
    size_t len = strlen(str) + 1;
    char *result = alloc_heap(len);
    memcpy(result, str, len);
    return result;
}

static char *alloc_qstring(const char *str)
{
    // Skip first and last quote
    str++;
    size_t len = strlen(str);
    char *result = alloc_heap(len);
    memcpy(result, str, len);
    result[len - 1] = '\0';
    return result;
}

struct term *term_new_number(int value)
{
    struct term *rv = alloc_heap(sizeof(struct term));
    rv->kind = term_number;
    rv->number = value;
    return rv;
}

struct term *term_new_string(const char *value)
{
    struct term *rv = alloc_heap(sizeof(struct term));
    rv->kind = term_string;
    rv->string = alloc_string(value);
    return rv;
}

struct term *term_new_qstring(const char *value)
{
    struct term *rv = alloc_heap(sizeof(struct term));
    rv->kind = term_string;
    rv->string = alloc_qstring(value);
    return rv;
}

struct term *term_new_boolean(bool value)
{
    struct term *rv = alloc_heap(sizeof(struct term));
    rv->kind = term_boolean;
    rv->boolean = value;
    return rv;
}

struct term *term_new_identifier(const char *value)
{
    struct term *rv = alloc_heap(sizeof(struct term));
    rv->kind = term_identifier;
    rv->identifier = alloc_string(value);
    return rv;
}

struct term *term_reverse(struct term *rv)
{
    struct term *prev = NULL;
    while (rv) {
        struct term *next = rv->next;
        rv->next = prev;
        prev = rv;
        rv = next;
    }
    return prev;
}

struct term *term_new_fun(const char *name, struct term *parameters)
{
    int arity = 0;
    struct term *p = parameters;
    while (p) {
        arity++;
        p = p->next;
    }

    fun_handler fun = lookup_function(name, arity);
    if (!fun)
        return NULL;

    struct term *rv = alloc_heap(sizeof(struct term));
    rv->kind = term_fun;
    rv->fun.fun = fun;
    rv->fun.parameters = parameters;
    return rv;
}

struct term *term_dupe(const struct term *rv)
{
    switch (rv->kind) {
    case term_identifier:
        return term_new_identifier(rv->identifier);
    case term_string:
        return term_new_string(rv->string);
    case term_number:
        return term_new_number(rv->number);
    case term_boolean:
        return term_new_boolean(rv->boolean);
    default:
        // Not supported
        return NULL;
    }
}

void inspect(const struct term *rv)
{
    switch (rv->kind) {
    case term_identifier:
        fprintf(stderr, "%s", rv->identifier);
        break;
    case term_string:
        fprintf(stderr, "\"%s\"", rv->string);
        break;
    case term_number:
        fprintf(stderr, "%d", rv->number);
        break;
    case term_boolean:
        fprintf(stderr, "%s", rv->boolean ? "true" : "false");
        break;
    case term_fun: {
        const struct term *p = rv->fun.parameters;
        const struct function_info *fun_info = function_info_by_fun(rv->fun.fun);
        fprintf(stderr, "%s(", fun_info->name);
        for (int i = 0; i < fun_info->arity; i++) {
            inspect(p);
            p = p->next;
            if (p)
                fprintf(stderr, ",");
        }
        fprintf(stderr, ")");
        break;
    }
    case term_variable:
        fprintf(stderr, "%s=", rv->var.name);
        inspect(rv->var.value);
        break;
    default:
        fprintf(stderr, "Unknown");
        break;
    }
}

static const struct term *run_function(const struct term *rv)
{
    assert(rv->kind == term_fun);

    if (rv->fun.fun)
        return rv->fun.fun(rv->fun.parameters);
    else
        return term_new_boolean(false);
}

const struct term *run_functions(const struct term *rv)
{
    const struct term *last_result = NULL;
    while (rv) {
        last_result = run_function(rv);
        rv = rv->next;
    }
    return last_result;
}

const struct term *term_resolve(const struct term *rv)
{
    switch (rv->kind) {
    case term_identifier:
        return term_resolve(get_variable(rv->identifier));
    case term_fun:
        return run_function(rv);
    default:
        return rv;
    }
}

/**
 * Compare two terms
 *
 * Returns:
 *   <0 if left < right
 *   0 if left == right
 *   >0 if left > right
 */
int term_compare(const struct term *left, const struct term *right)
{
    const struct term *rleft = term_resolve(left);
    const struct term *rright = term_resolve(right);

    // Make them the same...


    // Left and right are the same now.
    switch (rleft->kind) {
    case term_string:
        return strcmp(rleft->string, rright->string);
    case term_number:
        return rleft->number - rright->number;
    case term_boolean:
        // true > false
        return rleft->boolean - rright->boolean;
    default:
        return 1; // never happens
    }
}

bool term_to_boolean(const struct term *rv)
{
    rv = term_resolve(rv);
    switch (rv->kind) {
    case term_string:
        // empty strings, "0", and "false" are false. Everything else is true.
        if (strlen(rv->string) == 0 ||
            strcasecmp(rv->string, "0") == 0 ||
            strcasecmp(rv->string, "false") == 0)
            return false;
        else
            return true;
    case term_number:
        return rv->number != 0;
    case term_boolean:
        return rv->boolean;
    default:
        return false;
    }
}

int term_to_number(const struct term *rv)
{
    rv = term_resolve(rv);
    switch (rv->kind) {
    case term_string:
        return strtoull(rv->string, NULL, 0);
    case term_number:
        return rv->number;
    case term_boolean:
        return rv->boolean ? 1 : 0;
    default:
        return 0;
    }
}

const struct term *term_to_string(const struct term *rv)
{
    rv = term_resolve(rv);
    switch (rv->kind) {
    case term_identifier:
        return term_to_string(get_variable(rv->identifier));
    case term_string:
        return rv;
    case term_number:
    {
        char buffer[32];
        sprintf(buffer, "%d", rv->number);
        return term_new_string(buffer);
    }
    case term_boolean:
        return term_new_string(rv->boolean ? "true" : "false");
    default:
        return term_new_string("");
    }
}

static struct term *get_variable_impl(const char *name)
{
    for (struct term *v = variables; v; v = v->next) {
        if (strcmp(v->var.name, name) == 0)
            return v;
    }
    return NULL;
}

const struct term *get_variable(const char *name)
{
    struct term *var = get_variable_impl(name);
    if (var)
        return var->var.value;
    else
        return term_new_string("");
}

const char *get_variable_as_string(const char *name)
{
    struct term *var = get_variable_impl(name);
    if (var)
        return term_to_string(var->var.value)->string;
    else
        return "";
}

bool get_variable_as_boolean(const char *name)
{
    struct term *var = get_variable_impl(name);
    if (var)
        return term_to_boolean(var->var.value);
    else
        return false;
}
int get_variable_as_number(const char *name)
{
    struct term *var = get_variable_impl(name);
    if (var)
        return term_to_number(var->var.value);
    else
        return 0;
}

void set_variable(const char *name, const struct term *value)
{
    struct term *var = get_variable_impl(name);
    if (var) {
        var->var.value = value;
    } else {
        struct term *rv = alloc_heap(sizeof(struct term));
        rv->kind = term_variable;
        rv->var.name = alloc_string(name);
        rv->var.value = value;
        rv->next = variables;
        variables = rv;
    }
}

void set_string_variable(const char *name, const char *value)
{
    set_variable(name, term_new_string(value));
}

void set_boolean_variable(const char *name, bool value)
{
    set_variable(name, term_new_boolean(value));
}

void set_number_variable(const char *name, int value)
{
    set_variable(name, term_new_number(value));
}

static const struct term *function_assign(const struct term *parameters)
{
    const struct term *var = parameters;
    const struct term *value = term_resolve(parameters->next);

    set_variable(var->identifier, value);

    return value;
}

static const struct term *function_add(const struct term *parameters)
{
    int a = term_to_number(parameters);
    int b = term_to_number(parameters->next);
    return term_new_number(a + b);
}

static const struct term *function_subtract(const struct term *parameters)
{
    int a = term_to_number(parameters);
    int b = term_to_number(parameters->next);
    return term_new_number(a - b);
}

static const struct term *function_print(const struct term *parameters)
{
    while (parameters) {
        const struct term *str = term_to_string(parameters);
        fprintf(stderr, "%s", str->string);
        parameters = parameters->next;
    }
    fprintf(stderr, "\n");
    return NULL;
}

static const struct term *function_vars(const struct term *parameters)
{
    (void)parameters;

    for (const struct term *var = variables; var; var = var->next) {
        inspect(var);
        fprintf(stderr, "\n");
    }
    return NULL;
}

static const struct term *function_env(const struct term *parameters)
{
    (void)parameters;

    for (const struct uboot_name_value *var = working_uboot_env.vars; var; var = var->next) {
        fprintf(stderr, "%s=%s\n", var->name, var->value);
    }

    return NULL;
}
static const struct term *function_loadenv(const struct term *parameters)
{
    (void)parameters;
    const char *devpathspec = get_variable_as_string("uboot_env.path");
    char devpath[BLOCK_DEVICE_PATH_LEN];

    int block = get_variable_as_number("uboot_env.start");
    int block_count = get_variable_as_number("uboot_env.count");

    working_uboot_env.env_size = block_count * 512;
    int fd = open_block_device(devpathspec, O_RDONLY, devpath);
    if (fd < 0) {
        info("Could not open '%s'", devpathspec);
        return NULL;
    }
    if (lseek(fd, block * 512, SEEK_SET) < 0) {
        info("Could not seek to block %d (byte offset %d) in '%s'", block, block * 512, devpath);
        close(fd);
        return NULL;
    }
    char *buffer = malloc(working_uboot_env.env_size);
    ssize_t amount_read = read(fd, buffer, working_uboot_env.env_size);
    close(fd);

    if (amount_read != (ssize_t) working_uboot_env.env_size) {
        free(buffer);
        info("Could not read %d blocks (%d bytes) from '%s'", block_count, working_uboot_env.env_size, devpath);
        return NULL;
    }

    set_boolean_variable("uboot_env.loaded", false);
    set_boolean_variable("uboot_env.modified", false);

    if (uboot_env_read(&working_uboot_env, buffer) < 0) {
        free(buffer);
        return NULL;
    }
    free(buffer);

    set_boolean_variable("uboot_env.loaded", true);

    return NULL;
}
static const struct term *function_setenv(const struct term *parameters)
{
    const char *name = term_to_string(parameters)->string;
    const char *value = term_to_string(parameters->next)->string;

    if (uboot_env_setenv(&working_uboot_env, name, value) < 0) {
        info("Error setting uboot environment variable '%s' to '%s'", name, value);
    } else {
        set_boolean_variable("uboot_env.modified", true);
    }

    return parameters->next;
}
static const struct term *function_getenv(const struct term *parameters)
{
    const char *name = term_to_string(parameters)->string;

    char *value;
    if (uboot_env_getenv(&working_uboot_env, name, &value) < 0) {
        return term_new_string("");
    } else {
        struct term *ret = term_new_string(value);
        free(value);
        return ret;
    }
}
static const struct term *function_saveenv(const struct term *parameters)
{
    (void)parameters;

    const char *devpathspec = get_variable_as_string("uboot_env.path");
    char devpath[BLOCK_DEVICE_PATH_LEN];

    int block = get_variable_as_number("uboot_env.start");
    int block_count = get_variable_as_number("uboot_env.count");

    working_uboot_env.env_size = block_count * 512;
    int fd = open_block_device(devpathspec, O_WRONLY, devpath);
    if (fd < 0) {
        info("Could not open '%s'", devpathspec);
        return NULL;
    }
    if (lseek(fd, block * 512, SEEK_SET) < 0) {
        info("Could not seek to block %d (byte offset %d) in '%s'", block, block * 512, devpath);
        close(fd);
        return NULL;
    }
    char *buffer = malloc(working_uboot_env.env_size);
    if (uboot_env_write(&working_uboot_env, buffer) < 0) {
        free(buffer);
        return NULL;
    }

    ssize_t amount_written = write(fd, buffer, working_uboot_env.env_size);
    close(fd);
    free(buffer);

    if (amount_written != (ssize_t) working_uboot_env.env_size) {
        info("Could not write %d blocks (%d bytes) from '%s'", block_count, working_uboot_env.env_size, devpath);
        return NULL;
    }

    set_boolean_variable("uboot_env.modified", false);
    return NULL;
}
static const struct term *function_blkid(const struct term *parameters)
{
    (void)parameters;
    struct block_device_info *devices;

    probe_block_devices(&devices);

    for (struct block_device_info *device = devices; device; device = device->next) {
        fprintf(stderr, "%s: %sUUID=\"%s\"\n",
            device->path,
            device->type == BLOCK_DEVICE_DISK ? "DISK" : "PART",
            device->uuid);
    }

    free_block_devices(devices);
    return NULL;
}
static const struct term *function_cmd(const struct term *parameters)
{
    const int max_args = 15;
    char *argv[max_args + 1];
    int index = 0;
    while (parameters && index < max_args) {
        const struct term *str = term_to_string(parameters);
        argv[index] = str->string;
        parameters = parameters->next;
        index++;
    }
    argv[index] = 0;

    char output_buffer[256];
    output_buffer[0] = '\0';
    if (system_cmd(argv, output_buffer, sizeof(output_buffer)) != 0)
        info("Ignoring non-zero exit from %s", argv[0]);

    // Trim the output before returning since that's what's expected
    // in practice. There's usually a newline to trim anyway.
    trim_string_in_place(output_buffer);

    return term_new_string(output_buffer);
}
static const struct term *function_fwup_revert(const struct term *parameters)
{
    (void)parameters;

    const char *devpathspec = get_variable_as_string("uboot_env.path");
    char devpath[BLOCK_DEVICE_PATH_LEN];
    if (find_block_device_by_spec(devpathspec, devpath) < 0) {
        info("");
        return term_new_boolean(false);
    }

    char *const argv[7] = {"/usr/bin/fwup", "revert.fw", "-d", devpath, "-t", "revert", 0};

    char output_buffer[256];
    output_buffer[0] = '\0';
    if (system_cmd(argv, output_buffer, sizeof(output_buffer)) != 0) {
        info("Failure from fwup revert: %s", output_buffer);
        return term_new_boolean(false);
    }


    reboot(LINUX_REBOOT_CMD_RESTART);
    exit(EXIT_FAILURE);
}
static const struct term *function_ls(const struct term *parameters)
{
    const char *path = parameters ? term_to_string(parameters)->string : "/";

    struct dirent **namelist;
    int n = scandir(path,
                    &namelist,
                    NULL,
                    alphasort);
    int i;
    for (i = 0; i < n; i++) {
        fprintf(stderr, "%-32s%s\n",  namelist[i]->d_name, (namelist[i]->d_type == DT_DIR ? "<DIR>" : ""));
    }

    if (n >= 0) {
        for (i = 0; i < n; i++)
            free(namelist[i]);
        free(namelist);
    }
    return NULL;
}
static const struct term *function_reboot(const struct term *parameters)
{
    (void)parameters;
    reboot(LINUX_REBOOT_CMD_RESTART);
    exit(EXIT_FAILURE);
}
static const struct term *function_poweroff(const struct term *parameters)
{
    (void)parameters;
    reboot(LINUX_REBOOT_CMD_POWER_OFF);
    exit(EXIT_FAILURE);
}
static const struct term *function_sleep(const struct term *parameters)
{
    int milliseconds = term_to_number(parameters);
    usleep(milliseconds * 1000);
    return NULL;
}
static const struct term *function_readfile(const struct term *parameters)
{
    const char *path = term_to_string(parameters)->string;

    FILE *fp = fopen(path, "rb");
    if (fp) {
        char buffer[HEAP_SIZE / 4];
        size_t len = fread(buffer, 1, sizeof(buffer) - 1, fp);
        buffer[len] = 0;
        fclose(fp);
        return term_new_string(buffer);
    } else {
        info("Error reading %s", path);
        return term_new_string("");
    }
}

static const struct term *function_help(const struct term *parameters);

// Function lookup table
static struct function_info function_table[] = {
    {"=", 2, function_assign, NULL},
    {"+", 2, function_add, NULL},
    {"-", 2, function_subtract, NULL},
    {"blkid", 0, function_blkid, "list block devices"},
    {"cmd", 1, function_cmd, "run an external command"},
    {"env", 0, function_env, "print all loaded U-Boot variables"},
    {"fwup_revert", 0, function_fwup_revert, "revert to the previous firmware image"},
    {"getenv", 1, function_getenv, "get the value of a U-Boot variable"},
    {"help", 0, function_help, "print out help in the REPL"},
    {"loadenv", 0, function_loadenv, "load a U-Boot environment block"},
    {"ls", 0, function_ls, "list files"},
    {"poweroff", 0, function_poweroff, "power off the device"},
    {"print", 1, function_print, "print one or more strings and variables"},
    {"readfile", 1, function_readfile, "read a file (truncates long files)"},
    {"reboot", 0, function_reboot, "reset the device"},
    {"saveenv", 0, function_saveenv, "save all U-Boot variables back to storage"},
    {"setenv", 2, function_setenv, "set a U-Boot variable. It is not saved until you call saveenv/0"},
    {"sleep", 0, function_sleep, "sleep for n milliseconds"},
    {"vars", 0, function_vars, "print all known variables and their values"},
    {NULL, 0, NULL, NULL}
};

fun_handler lookup_function(const char *name, int arity)
{
    struct function_info *entry = function_table;
    while (entry->name) {
        if (strcmp(entry->name, name) == 0 && arity >= entry->arity)
            return entry->handler;
        entry++;
    }
    return NULL;
}

const struct function_info *function_info_by_fun(fun_handler fun)
{
    struct function_info *entry = function_table;
    while (entry->handler && entry->handler != fun)
        entry++;

    return entry;
}

const struct term *function_help(const struct term *parameters)
{
    (void)parameters;

    struct function_info *entry = function_table;
    while (entry->handler) {
        if (entry->description) {
            char buf[16];
            snprintf(buf, sizeof(buf), "%s/%d", entry->name, entry->arity);
            fprintf(stderr, "%-16s%s\n", buf, entry->description);
        }
        entry++;
    }

    fprintf(stderr, "\nType Ctrl-D to exit repl.\n");
    return NULL;
}

int lexer_set_input(char *input);
int lexer_set_file(const char *path);
int yyget_lineno(void);
int yyparse(void);

int eval_string(char *input)
{
    term_gc_heap();
    lexer_set_input(input);
    return yyparse();
}

int eval_file(const char *path)
{
    term_gc_heap();
    if (lexer_set_file(path) < 0)
      return -1;

    return yyparse();
}

int yyerror(char const *msg)
{
  fprintf(stderr, "Error on line %d: %s\n", yyget_lineno(), msg);
  return 0;
}
